AsyncDisplayKit/ASDisplayNode.mm

```diff
@@ -703,6 +704,69 @@ - (instancetype)styledWithBlock:(void (^)(ASLayoutElementStyle *style))styleBloc
 
 #pragma mark - Layout
 
+- (void)invalidateSize
+{
+  ASDisplayNodeAssertThreadAffinity(self);
+  
+  __instanceLock__.lock();
+
+  // Mark the node for layout in the next layout pass
+  [self setNeedsLayout];
+  
+  ASDisplayNode *supernode = _supernode;
+  if (supernode) {
+    __instanceLock__.unlock();
+    // Cause supernode's layout to be invalidated
+    // We need to release the lock to prevent a deadlock
+    [supernode invalidateSize];
+    return;
+  }
+  
+  // We are the root node and need to re-flow the layout; one of our children requested to have its size re-set.
+  CGSize oldSize = self.bounds.size;
+
+  // Figure out constrainedSize to use
+  ASSizeRange constrainedSize = ASSizeRangeMake(self.bounds.size);
+  if (_pendingDisplayNodeLayout != nullptr) {
+    constrainedSize = _pendingDisplayNodeLayout->constrainedSize;
+  } else if (_calculatedDisplayNodeLayout->layout != nil) {
+    constrainedSize = _calculatedDisplayNodeLayout->constrainedSize;
+  }
+
+  // Check if the returned layout has a different size as the current bounds
+  ASLayout *layout = [self layoutThatFits:constrainedSize];
```

Comment:
It's important to differentiate between layout and measure pass here. calling `layoutThatFits:` just performs a measure pass and no layout pass immediately. If we would call here `layoutIfNeeded` it would do a layout pass and and go into `__layout` what could happen in there that the size of the node changed and would call again `invalidateSize` and a loop will occur. That's the reason we only should do a measure pass in here in this case.